vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -537,6 +575,83 @@ float ConnectionBoxMapLookahead::get_expected_cost(
     }
 }
 
+static int manhattan_distance(const t_rr_node& node, int x, int y) {
+    int node_center_x = (node.xhigh() + node.xlow()) / 2;
+    int node_center_y = (node.yhigh() + node.ylow()) / 2;
+    return abs(node_center_x - x) + abs(node_center_y - y);
+}
+
+static BoundingBox<int> bounding_box_for_node(const t_rr_node& node) {
+    return BoundingBox<int>(node.xlow(), node.ylow(),
+                            node.xhigh(), node.yhigh());
+}
+
+void ConnectionBoxMapLookahead::find_start_inodes_for_segment_types(int num_segments) {
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& rr_nodes = device_ctx.rr_nodes;
+
+    // compute bounding boxes for each segment type
+    bounding_box_for_segment_.clear();
+    bounding_box_for_segment_.resize(num_segments, BoundingBox<int>());
+    for (size_t i = 0; i < rr_nodes.size(); i++) {
+        auto& node = rr_nodes[i];
+        if (node.type() != CHANX && node.type() != CHANY) continue;
+        int seg_index = device_ctx.rr_indexed_data[node.cost_index()].seg_index;
+
+        VTR_ASSERT(seg_index != OPEN);
+        VTR_ASSERT(seg_index < num_segments);
+
+        bounding_box_for_segment_[seg_index] |= bounding_box_for_node(node);
+    }
+
+    // select an inode near the center of the bounding box for each segment type
+    start_inode_for_segment_.clear();
+    start_inode_for_segment_.resize(num_segments);
+    for (auto& grid : start_inode_for_segment_) {
+        for (auto& row : grid) {
+            for (auto& cell : row) {
+                cell = std::list<ssize_t>();
+            }
+        }
+    }
+    for (size_t i = 0; i < rr_nodes.size(); i++) {
```

Comment:
Is a 6 layer nested loop with a goto really the clearest this code can be made?

Reply:
There's only three loops, over the three dimensions, the node, x & y. `goto` is the cleanest way to continue to the outer loop; when a position is found for a node, move on to the next.  Maybe a better label name will help.

Otherwise, I'd have to use a "found" boolean, which seems more error prone to me because you have to remember to reset it, check it, and set it, and it would effectively to the same thing as a single `goto`.